### PR TITLE
Merge improved autocomplete

### LINKED ISSUE #379: Inline suggestions for VSCoq 2
As part of our Master's thesis, @Hjaltesorgenfrei, @Jakobis and I want to improve Coq with inline suggestions/completion items.

The idea is to bring Coq "up to standard" with other mainstream/general-purpose programming languages in terms of providing suggestions for lemmas, tactics, goal hypotheses, etc., and also filter and sort the suggestions such that they are relevant to the current context. 

This issue will be long-standing as we work on different aspects of providing these suggestions, and we intend to make the changes in shorter-lived PRs and issues. 
We will link this issue with other issues and PRs we submit, so we can keep a thread over the larger narrative.

Currently, as described in #377, we have some suggestions already implemented. 

Below I have attached an image of our current prototype, in case it is not obvious what feature we are implementing:
image

### PR DESCRIPTION
**Pull request by Duckapple**
Here is the type-aware code completion we (@Hjaltesorgenfrei, @Jakobis and I) have made for our Master's thesis [which can be read here](https://github.com/Jakobis/vscoqComparison/blob/main/Expanding_Coq_with_Type_Aware_Code_Completion.pdf).
While we implemented 5 algorithms for our thesis, through benchmarking we found out there is not that big of a reason for keeping all of them, as they perform very similarly, and the user wouldn't know which one to use anyways. We kept the two which work best for `apply` and `rewrite` respectively, such that we can potentially improve completion by switching to the better one of them depending on what is already written.
The benchmarking was done in this repository: https://github.com/Jakobis/vscoqComparison

Closes #379

### CODE REVIEW

File: language-server/dm/executionManager.ml [hunk 1]
```diff
@@ -569,6 +582,15 @@ let get_context st id =
     let sigma = Evd.from_env env in
     Some (sigma, env)
 
+let get_completions st pos =
+  unfreeze_interp_state st pos;
```
Comment (Hjaltesorgenfrei) [inline language-server/dm/executionManager.ml:586]: Should the state be frozen again afterwards?

File: language-server/dm/executionManager.ml [hunk 2]
```diff
@@ -569,6 +581,33 @@ let get_context st id =
...
+      None
+    | _ -> log "Cannot unfreeze state";
+    None   
+*)
+
+let get_completions st id =
+  let aux () = 
+    match find_fulfilled_opt id st.of_sentence with
+    | Some (Success (Some { interp = interp_state; synterp = synterp_state })) when Option.has_some interp_state.lemmas ->
+      Vernacstate.Interp.unfreeze_interp_state interp_state;
```
Comment (maximedenes) [inline language-server/dm/executionManager.ml:599]: Why do you need to imperatively unfreeze the system state here?

File: language-server/dm/executionManager.ml [hunk 3]
```diff
@@ -569,6 +581,33 @@ let get_context st id =
     let sigma = Evd.from_env env in
     Some (sigma, env)
 
+(*
+    match find_fulfilled_opt id st.of_sentence with
```
Comment (maximedenes) [inline language-server/dm/executionManager.ml:585]: Leftover code

File: language-server/dm/executionManager.ml (hunk 2 above)
Comment (Hjaltesorgenfrei) [reply language-server/dm/executionManager.ml:599]: Because of this issue https://github.com/coq-community/vscoq/issues/458

**Review by maximedenes — APPROVED** [review]

**Pull request merged**